integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/Spark3EventHandlerFactory.java

```diff
@@ -0,0 +1,57 @@
+package io.openlineage.spark3.agent.lifecycle.plan;
+
+import com.google.common.collect.ImmutableList;
+import io.openlineage.client.OpenLineage;
+import io.openlineage.client.OpenLineage.DatasetFacet;
+import io.openlineage.client.OpenLineage.DatasetVersionDatasetFacet;
+import io.openlineage.client.OpenLineage.OutputDatasetFacet;
+import io.openlineage.spark.api.CustomFacetBuilder;
+import io.openlineage.spark.api.OpenLineageContext;
+import io.openlineage.spark.api.OpenLineageEventHandlerFactory;
+import io.openlineage.spark3.agent.facets.builder.DatasetVersionOutputDatasetFacetBuilder;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.BiConsumer;
+import lombok.AllArgsConstructor;
+import org.apache.spark.package$;
+
+public class Spark3EventHandlerFactory implements OpenLineageEventHandlerFactory {
+  @Override
+  public List<CustomFacetBuilder<?, ? extends OutputDatasetFacet>> createOutputDatasetFacetBuilders(
+      OpenLineageContext context) {
+    if (package$.MODULE$.SPARK_VERSION().startsWith("2")) {
+      return Collections.emptyList();
+    }
+    return ImmutableList.of(
+        new DatasetVersionFacetConverter<>(
+            context, new DatasetVersionOutputDatasetFacetBuilder(context)));
+  }
+
+  @AllArgsConstructor
+  private static class DatasetVersionFacetConverter<T>
```

Comment:
Because I wanted to 

1.  have `DatasetVersionDatasetFacet` as a core facet that's attachable to both input and output datasets. If we did it the other way, I'd have to have two exactly same facets in core spec. It could be solved by making `OutputDatasetFacet` and `InputDatasetFacet` extend `DatasetFacet`.
2. Original version of this PR had also `DatasetVersionInputDatasetFacetBuilder` that utilized the same converter feature.

Reply:
I submitted [this PR](https://github.com/OpenLineage/OpenLineage/pull/513), which I think will help. If we can make the `DatasetVersionDatasetFacet` implement both the `InputDatasetFacet` and the `OutputDatasetFacet`, I think we can simplify this code so that only a single facet builder is needed.